Fixed Auto-dimming effect in portal window of APMP playback from WebView's Div Element Fullscreen

Phinny01 opened the pull request:
#### 52379892b081a3a346c979641831dedddcf224c7
<pre>
Fixed Auto-dimming effect in portal window of APMP playback from WebView&apos;s Div Element Fullscreen
<a href="https://bugs.webkit.org/show_bug.cgi?id=296495">https://bugs.webkit.org/show_bug.cgi?id=296495</a>
<a href="https://rdar.apple.com/154223902">rdar://154223902</a>

Reviewed by Andy Estes and Jean-Yves Avenard.

This patch creates a single place to store a user preference: whether videos should automatically
dim the background when played in fullscreen. PlaybackSessionModel carries a boolean
m_prefersAutoDimming, with accessors that notify its clients when the value changes.
PlaybackSessionModelContext forwards the change to VideoPresentationManagerProxy, allowing every
concurrent playback context to observe the current setting. PlaybackSessionManagerProxy persists
the value to user defaults and relays it to the Web-process fullscreen manager so the system knows
which darkness level to apply. VideoPresentationInterfaceLMK applies the flag to the underlying
AVPlayerViewController, caches the previous stage darkness on entry and restores it on exit.
WKFullScreenWindowControllerIOS consults the proxy when entering fullscreen, reflects user changes
made regarding autoDimming back into the proxy.

* Source/WebCore/platform/cocoa/PlaybackSessionModel.h:
(WebCore::PlaybackSessionModel::prefersAutoDimming const):
(WebCore::PlaybackSessionModel::setPrefersAutoDimming):
* Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.mm:
(WebKit::VideoPresentationInterfaceLMK::presentFullscreen):
(WebKit::VideoPresentationInterfaceLMK::dismissFullscreen):
(WebKit::VideoPresentationInterfaceLMK::ensurePlayableViewController):
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.mm:
(WebKit::PlaybackSessionModelContext::PlaybackSessionModelContext):
(WebKit::PlaybackSessionModelContext::setPrefersAutoDimming):
(WebKit::PlaybackSessionManagerProxy::prefersAutoDimming const):
(WebKit::PlaybackSessionManagerProxy::setPrefersAutoDimming):
* Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.mm:
(WebKit::VideoPresentationManagerProxy::didExitFullscreen):
(WebKit::VideoPresentationManagerProxy::didEnterFullscreen):
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/WebFullScreenManagerProxy.h:
(WebKit::WebFullScreenManagerProxy::client):
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::fullScreenManagerProxyClient):
(WebKit::PageClientImpl::didEnterFullscreen):
(WebKit::PageClientImpl::didExitFullscreen):
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.h:
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
(-[WKFullScreenWindowController initWithWebView:]):
(-[WKFullScreenWindowController didEnterVideoFullscreen]):
(-[WKFullScreenWindowController didExitVideoFullscreen]):
(-[WKFullScreenWindowController prefersSceneDimming]):
(-[WKFullScreenWindowController _performSpatialFullScreenTransition:completionHandler:]):
(-[WKFullScreenWindowController toggleSceneDimming]):
* Source/WebKit/WebKitSwift/LinearMediaKit/LinearMediaTypes.swift:
(WKSPlayableViewControllerHost.prefersAutoDimming):
(WKSPlayableViewControllerHost.playable):
* Source/WebKit/WebKitSwift/LinearMediaKit/WKSLinearMediaTypes.h:

Canonical link: <a href="https://commits.webkit.org/298720@main">https://commits.webkit.org/298720@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ccba76bdea583325fa4fa8124479d1ae4de0b6c7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/116419 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/36080 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/26631 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/122476 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/66982 "Built successfully") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/118308 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/36774 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/44668 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/88418 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42896 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/119368 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/29324 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/104442 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/68857 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/28404 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/22546 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/66157 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/98706 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/22707 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/125625 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/43313 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/32525 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/97124 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/43678 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/100649 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/96919 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24668 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/42212 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/20114 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/39267 "The change is no longer eligible for processing.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/43201 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/48792 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/143645 "Build is in progress. Recent messages:") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/42667 "Built successfully") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/35/builds/143645 "Build is in progress. Recent messages:") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/46007 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/44372 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebKit/WebKitSwift/LinearMediaKit/LinearMediaTypes.swift:
```diff
@@ -96,26 +96,33 @@
-    @nonobjc final var prefersAutoDimming: Bool {
+    @nonobjc
+    final var playable: (any Playable)?
+    {
```
this brace should be on the preceding line

aestes commented on Source/WebCore/platform/cocoa/VideoPresentationModel.h:
```diff
@@ -113,6 +116,8 @@ class VideoPresentationModel
     virtual const Logger* loggerPtr() const { return nullptr; }
 #endif
+private:
+    bool m_prefersAutoDimming { true };
```
Nit: the default state should be false

aestes commented on Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.h:
```diff
@@ -30,6 +30,10 @@
 #include <WebCore/VideoPresentationInterfaceIOS.h>
 #include <wtf/TZoneMalloc.h>
 
+#if PLATFORM(VISION)
```
The `#if` isn't necessary

aprotyas commented on Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.h:
```diff
@@ -140,6 +143,7 @@ class VideoPresentationModelContext final
     WebCore::FloatSize m_videoDimensions;
     bool m_hasVideo { false };
     bool m_isChildOfElementFullscreen { false };
+    bool m_prefersAutoDimming { true };
```
nit: you could use `m_prefersAutoDimming` from `VideoPresentationModel` if it's a protected member?

aprotyas replied:
(prevents creating yet another `m_prefersAutoDimming`)

Phinny01 replied:
Right. I have addressed now

aprotyas replied:
Hmm this is still here?

Phinny01 replied:
Removed it now

aprotyas commented on Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.mm:
```diff
@@ -1182,6 +1236,15 @@ - (BOOL)prefersStatusBarHidden
+                controllerClient->setPrefersAutoDimming(model->prefersAutoDimming());
+        }
+    }
+#endif
```
You could consider reusing this block between `[enter|exit]Fullscreen`

Phinny01 replied:
I made a helper to use between the two methods

aprotyas commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -872,6 +880,39 @@ - (void)resetSupportedOrientations
+    if (auto* videoPresentationManager = self._videoPresentationManager) {
+        videoPresentationManager->setPrefersAutoDimming(prefersAutoDimming);
+
+        if (RefPtr lmk = videoPresentationManager->controlsManagerInterface())
```
Maybe a different name, `interface`?

Phinny01 replied:
I have addressed  this

aprotyas commented on Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.h:
```diff
@@ -96,6 +101,9 @@ class VideoPresentationInterfaceLMK final : public WebCore::VideoPresentationInt
     CompletionHandler<void(bool)> m_exitExternalPlaybackHandler;
+
+    MRUIDarknessPreference m_savedDarknessPreference;
+    BOOL m_savedAutoDimmingPreference;
```
You should initialize these here instead of in the constructor.


```suggestion
    MRUIDarknessPreference m_savedDarknessPreference { MRUIDarknessPreferenceUnspecified };
    BOOL m_savedAutoDimmingPreference { YES };
```

Phinny01 replied:
fixed this

aprotyas commented on Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.mm:
```diff
@@ -85,7 +87,11 @@ - (void)layoutSublayers
 VideoPresentationInterfaceLMK::VideoPresentationInterfaceLMK(WebCore::PlaybackSessionInterfaceIOS& playbackSessionInterface)
     : VideoPresentationInterfaceIOS { playbackSessionInterface }
+    , m_savedDarknessPreference { MRUIDarknessPreferenceUnspecified }
+    , m_savedAutoDimmingPreference { YES }
```
No longer needed if you initialize these during member declaration.

aprotyas commented on Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.mm:
```diff
@@ -136,18 +142,51 @@ - (void)layoutSublayers
+            if (RefPtr model = this->videoPresentationModel())
+                m_savedAutoDimmingPreference = model->prefersAutoDimming();
+
+            playableViewController().prefersAutoDimming = m_savedAutoDimmingPreference;
```
Since you're always going to prefer `model->prefersAutoDimming()`, why not set `playableViewController().prefersAutoDimming` conditionally and read from it unconditionally.


```suggestion
            if (RefPtr model = this->videoPresentationModel())
                playableViewController().prefersAutoDimming = model->prefersAutoDimming();
            m_savedAutoDimmingPreference = playableViewController().prefersAutoDimming;
```

Phinny01 replied:
I have done that now

aprotyas commented on Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.mm:
```diff
@@ -322,6 +322,36 @@ - (BOOL)prefersStatusBarHidden
+
+void VideoPresentationModelContext::setPrefersAutoDimming(bool prefersAutoDimming)
+{
+    if (VideoPresentationModel::prefersAutoDimming() == prefersAutoDimming)
```
You're calling the parent implementation anyway, but let that be a detail for the getter.

```suggestion
    if (prefersAutoDimming() == prefersAutoDimming)
```

aprotyas commented on Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.mm:
```diff
@@ -538,6 +568,25 @@ - (BOOL)prefersStatusBarHidden
+
+void VideoPresentationManagerProxy::setPrefersAutoDimming(bool prefersAutoDimming)
+{
+    if (this->prefersAutoDimming()  == prefersAutoDimming)
```
Extra whitespace

```suggestion
    if (this->prefersAutoDimming() == prefersAutoDimming)
```

aprotyas commented on Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.mm:
```diff
@@ -1157,6 +1206,10 @@ - (BOOL)prefersStatusBarHidden
 
+#if ENABLE(FULLSCREEN_API)
+    if (RefPtr page = m_page.get())
+        VideoPresentationManagerProxy::updateAutoDimming(contextId, page);
```
Why the class scope?
```suggestion
        updateAutoDimming(contextId, page);
```

aprotyas commented on Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.mm:
```diff
@@ -1182,6 +1246,9 @@ - (BOOL)prefersStatusBarHidden
         return;
     }
+#if ENABLE(FULLSCREEN_API)
+    VideoPresentationManagerProxy::updateAutoDimming(contextId, page);
```
```suggestion
    updateAutoDimming(contextId, page);
```

aprotyas commented on Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.mm:
```diff
@@ -322,6 +322,36 @@ - (BOOL)prefersStatusBarHidden
+        manager->setPrefersAutoDimming(prefersAutoDimming);
+
+    ALWAYS_LOG_IF_POSSIBLE(LOGIDENTIFIER, prefersAutoDimming);
+    m_clients.forEach([&](auto& client) {
```
Nit: you can just capture `prefersAutoDimming`.


```suggestion
    m_clients.forEach([prefersAutoDimming](auto& client) {
```

aprotyas commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -811,7 +814,12 @@ - (id)initWithWebView:(WKWebView *)webView
 #endif
-
+#if PLATFORM(VISION)
+    if (auto* videoPresentationManager = [self _videoPresentationManager])
```
To align with our SaferCPP guidelines.
```suggestion
    if (RefPtr videoPresentationManager = [self _videoPresentationManager])
```

aprotyas commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -872,6 +880,39 @@ - (void)resetSupportedOrientations
+    }
+#endif
+
+    if (auto* videoPresentationManager = self._videoPresentationManager) {
```
```suggestion
    if (RefPtr videoPresentationManager = self._videoPresentationManager) {
```

aprotyas commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -1849,6 +1935,8 @@ - (BOOL)_sceneAspectRatioLockingEnabled
 
 - (BOOL)prefersSceneDimming
 {
+    if (auto* videoPresentationManager = [self _videoPresentationManager])
```
```suggestion
    if (RefPtr videoPresentationManager = [self _videoPresentationManager])
```

aprotyas commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -1384,6 +1429,40 @@ - (void)_reinsertWebViewUnderPlaceholder
 
+- (void)syncAutoDimmingStateWithVideoPresentationManager
+{
+    auto *videoPresentationManager = self._videoPresentationManager;
```
```suggestion
    RefPtr videoPresentationManager = self._videoPresentationManager;
```

aprotyas commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -2021,14 +2115,21 @@ - (void)toggleSceneDimming
     BOOL updatedPrefersSceneDimming = !self.prefersSceneDimming;
 
-    [[NSUserDefaults standardUserDefaults] setBool:updatedPrefersSceneDimming forKey:kPrefersFullScreenDimmingKey];
+    if (auto *videoPresentationManager = self._videoPresentationManager)
```
```suggestion
    if (RefPtr videoPresentationManager = self._videoPresentationManager)
```

aprotyas commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -1384,6 +1429,40 @@ - (void)_reinsertWebViewUnderPlaceholder
+    videoPresentationManager->setPrefersAutoDimming(globalPref);
+#endif
+
+#if PLATFORM(VISION)
```
Condense
```suggestion
```

aprotyas reviewed (commented):
@Phinny01 mostly style comments!

I think a previous comment about not reintroducing m_prefersAutoDimming in a child class may still need addressing?

Also, it would be great if you could also get a review from someone more familiar with the media code. The plumbing looks good to me but I don't know about correctness 🤷

jyavenard commented on Source/WebCore/platform/cocoa/VideoPresentationModel.h:
```diff
@@ -137,6 +144,9 @@ class VideoPresentationModelClient : public CanMakeWeakPtr<VideoPresentationMode
     virtual void audioSessionCategoryChanged(AudioSessionCategory, AudioSessionMode, RouteSharingPolicy) { }
     virtual void hasBeenInteractedWith() { }
+    virtual void prefersAutoDimmingChanged(bool) { }
+    virtual void willPresentFullscreen() = 0;
```
why makes those two methods pure virtual when for all implementations but one they do nothing ?

Phinny01 replied:
Did you mean virtual void `willDismissFullscreen() = 0;` instead of `virtual void prefersAutoDimmingChanged(bool) { }`

jyavenard replied:
` T method() = 0` is pure-virtual, it requires an implementation.

All your implementations but one do absolutely nothing. so `virtual void willPresentFullscreen() { }` ; that's if we needed that method though, I don't believe it's needed.

Phinny01 replied:
Yes I have changed those now

jyavenard commented on Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.h:
```diff
@@ -30,6 +30,10 @@
 #include <wtf/TZoneMalloc.h>
 
+#if PLATFORM(VISION)
+#import "MRUIKitSPI.h"
```
what is this header for? it doesn't appear to be used in that header.

If needed, shouldn't this be added to the code file instead?

jyavenard commented on Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.mm:
```diff
@@ -136,18 +140,43 @@ - (void)layoutSublayers
         if (auto* playbackSessionModel = this->playbackSessionModel()) {
             playbackSessionModel->setSpatialTrackingLabel(m_spatialTrackingLabel);
             playbackSessionModel->setSoundStageSize(WebCore::AudioSessionSoundStageSize::Large);
+            m_savedAutoDimmingPreference = playableViewController().prefersAutoDimming;
```
why set it twice, before and after setting prefersAutoDimming.

jyavenard commented on Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.mm:
```diff
@@ -136,18 +140,43 @@ - (void)layoutSublayers
+
+    if (RetainPtr playerController = m_playerViewController) {
+        [playerViewController setPrefersAutoDimming:prefersAutoDimming];
+        m_savedAutoDimmingPreference = [playerViewController prefersAutoDimming];
```
can the value changed between the time it is set and read again?

what about just:
```suggested
    if (std::exchange(m_savedAutoDimmingPreference, prefersAutoDimming) != prefersAutoDimming)
        [m_playerViewController setPrefersAutoDimming:prefersAutoDimming];
```

there's no need to test for nil with objC objects.

jyavenard commented on Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.mm:
```diff
@@ -749,18 +815,18 @@ - (BOOL)prefersStatusBarHidden
 }
 
-const VideoPresentationManagerProxy::ModelInterfacePair* VideoPresentationManagerProxy::findModelAndInterface(PlaybackSessionContextIdentifier contextId) const
+std::optional<std::reference_wrapper<const VideoPresentationManagerProxy::ModelInterfacePair>> VideoPresentationManagerProxy::findModelAndInterface(PlaybackSessionContextIdentifier contextId) const
```
Wrapping a reference in a std::optional adds no benefit to returning a pointer. It's no safer, just harder to use and IMHO less readable. this is an unnecessary change

jyavenard commented on Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.mm:
```diff
@@ -298,13 +327,19 @@ - (void)layoutSublayers
     ALWAYS_LOG_IF_POSSIBLE(LOGIDENTIFIER);
     m_playerViewController = [linearMediaPlayer() makeViewController];
-    [m_playerViewController viewController].view.alpha = 0;
+    if (m_playerViewController) {
```
with obj-C, there's no need to test if the pointer is nil.
and how m_playerViewController be nil all of a sudden? the method is non fallible.

jyavenard commented on Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.mm:
```diff
@@ -173,7 +202,7 @@ - (void)layoutSublayers
         }
 
-        handler(success, [m_playerViewController viewController]);
+        handler(success, [m_playerViewController.get() viewController]);
```
why this change? this isn't necessary, there's no need to call .get(), RetainPtr has an operator to allow for automatic conversion.

Phinny01 replied:
I have reverted it now

jyavenard replied:
Please don't resolve unless it's actually resolved and you've updated the new patch.

Phinny01 replied:
I have updated the new patch

jyavenard reviewed (commented):
This seems to be more complicated than it should be, and adding more methods that are doing the same thing that already exists (willEnterFullScreen / didExitFullscreen).
Can't you re-use the existing methods without adding new ones notifying the same?

Should store the attributes in the PlaybackSessionModel like all the other attributes. The VideoPresentationModelContext has access to the PlaybackSessionModel

jyavenard commented on Source/WebCore/platform/cocoa/WebAVPlayerLayer.mm:
```diff
@@ -65,6 +65,8 @@ @interface WebAVPlayerLayer (Logging)
     {
     }
+    void willPresentFullscreen() final { }
+    void willDismissFullscreen() final { }
```
like here, all your implementations are the same, so why make it pure virtual ?

Phinny01 replied:
I have changed that now. I have posted the patch

jyavenard commented on Source/WebCore/platform/cocoa/PlaybackSessionModel.h:
```diff
@@ -150,10 +150,16 @@ class PlaybackSessionModel : public CanMakeWeakPtr<PlaybackSessionModel> {
 #endif
+
+protected:
+    bool m_prefersAutoDimming { true };
```
let's not have this here considering it's only use on VisionOS,

Phinny01 replied:
I moved it to the PlaybackSessionManagerProxy.h instead as private

jyavenard commented on Source/WebCore/platform/cocoa/PlaybackSessionModel.h:
```diff
@@ -150,10 +150,16 @@ class PlaybackSessionModel : public CanMakeWeakPtr<PlaybackSessionModel> {
     virtual bool supportsLinearMediaPlayer() const { return false; }
 #endif
 
+    virtual bool prefersAutoDimming() const { return m_prefersAutoDimming; }
```
have the default return true; rather than having the need for protected variable. We typically don't want those here ; they should be private

Phinny01 replied:
I fixed that

jyavenard commented on Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.mm:
```diff
@@ -133,21 +137,39 @@ - (void)layoutSublayers
     playbackSessionInterface().stopObservingNowPlayingMetadata();
     [linearMediaPlayer() exitFullscreenWithCompletionHandler:makeBlockPtr([this, protectedThis = Ref { *this }, completionHandler = WTFMove(completionHandler)] (BOOL success, NSError *error) {
-        if (auto* playbackSessionModel = this->playbackSessionModel()) {
+        if (CheckedPtr playbackSessionModel = playbackSessionInterface().playbackSessionModel()) {
```
I think it would have been better to do all those changes in a separate patch, so that we narrow the scope to just auto dimming.

This isn't necessary for this change

Phinny01 replied:
I reverted that

jyavenard commented on Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.mm:
```diff
@@ -169,7 +191,7 @@ - (void)layoutSublayers
     // Puts the player into `enteringExternal` state.
     [linearMediaPlayer() enterExternalPlaybackWithCompletionHandler:makeBlockPtr([this, protectedThis = Ref { *this }, handler = WTFMove(enterHandler)] (BOOL success, NSError *error) mutable {
-        if (auto* playbackSessionModel = this->playbackSessionModel()) {
+        if (CheckedPtr playbackSessionModel = playbackSessionInterface().playbackSessionModel()) {
```
ditto, should be done in a related change, not here

jyavenard commented on Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.mm:
```diff
@@ -318,13 +340,20 @@ - (void)layoutSublayers
     ALWAYS_LOG_IF_POSSIBLE(LOGIDENTIFIER);
     m_playerViewController = [linearMediaPlayer() makeViewController];
-    [m_playerViewController viewController].view.alpha = 0;
+    [m_playerViewController.get() viewController].view.alpha = 0;
```
that's an unnecessary change, it's less slightly too IMHO.

jyavenard commented on Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.mm:
```diff
@@ -318,13 +340,20 @@ - (void)layoutSublayers
-    [m_playerViewController viewController].view.alpha = 0;
+    [m_playerViewController.get() viewController].view.alpha = 0;
+
+    if (RefPtr model = videoPresentationModel()) {
```
why this test when `model` isn't even used in the block?

jyavenard commented on Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.mm:
```diff
@@ -318,13 +340,20 @@ - (void)layoutSublayers
 {
     auto& otherInterface = static_cast<VideoPresentationInterfaceLMK&>(otherInterfaceIOS);
     std::swap(m_playerViewController, otherInterface.m_playerViewController);
+    std::swap(m_savedAutoDimmingPreference, otherInterface.m_savedAutoDimmingPreference);
```
as discussed, I don't think we want to use std::swap here, should just copy the value

Phinny01 replied:
Yes I changed that and just copied the value. that works

jyavenard commented on Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.h:
```diff
@@ -189,6 +189,9 @@ class PlaybackSessionModelContext final
     WebCore::AudioSessionSoundStageSize soundStageSize() const final { return m_soundStageSize; }
     void setSoundStageSize(WebCore::AudioSessionSoundStageSize) final;
 
+    bool prefersAutoDimming() const final { return PlaybackSessionModel::prefersAutoDimming(); }
```
don't use the base class to store data. As it is that override is totally unnecessary as that's what would happen by default anyway.

jyavenard commented on Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.mm:
```diff
@@ -1121,6 +1121,42 @@
+    if (!m_controlsManagerContextId)
+        return true;
+
+    auto it = m_contextMap.find(*m_controlsManagerContextId);
```
I see this pattern of searching for `m_controlsManagerContextId` a lot in this file, could be replaced with a helper. Would be a good follow up.

Phinny01 replied:
I will do that after this pr

jyavenard commented on Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.mm:
```diff
@@ -750,9 +750,9 @@ - (BOOL)prefersStatusBarHidden
-    if (it == m_contextMap.end())
-        return nullptr;
-    return &(it->value);
+        if (it == m_contextMap.end())
```
why this change? it's wrong too.

jyavenard reviewed (commented):
LGTM, some nits.

And you've deleted the WebCore project file, that ain't going to compile

jyavenard commented on Source/WebCore/WebCore.xcodeproj/project.pbxproj:
```diff
@@ -48002,13 +48002,27 @@
 			isa = XCBuildConfiguration;
 			baseConfigurationReference = 1CDD45E50BA9C84600F90147 /* WebCore.xcconfig */;
 			buildSettings = {
+				CODE_SIGN_IDENTITY = "Apple Development";
```
this needs to be reverted. Don't push that.
There's no need to modify the project.pbxproj

Always check your PR after uploading or double-check the commit to make sure you don't get stray files included.

jyavenard commented on Source/WebCore/platform/cocoa/WebAVPlayerLayer.mm:
```diff
@@ -65,6 +65,7 @@ @interface WebAVPlayerLayer (Logging)
         : m_playerLayer(playerLayer)
     {
     }
+
```
unrelated, shouldn't touch this file

jyavenard commented on Source/WebCore/platform/ios/VideoPresentationInterfaceAVKit.h:
```diff
@@ -38,7 +38,6 @@ class VideoPresentationInterfaceAVKit final : public VideoPresentationInterfaceI
 public:
     WEBCORE_EXPORT static Ref<VideoPresentationInterfaceAVKit> create(PlaybackSessionInterfaceIOS&);
     ~VideoPresentationInterfaceAVKit();
-
```
unrelated, shouldn't touch this file

jyavenard commented on Source/WebCore/platform/ios/VideoPresentationInterfaceTVOS.h:
```diff
@@ -43,7 +43,7 @@ class VideoPresentationInterfaceTVOS final : public VideoPresentationInterfaceIO
     bool isPlayingVideoInEnhancedFullscreen() const final { return false; }
     bool pictureInPictureWasStartedWhenEnteringBackground() const final { return false; }
-    
+
```
ditto. only commit files that actually modified.

jyavenard commented on Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.mm:
```diff
@@ -133,21 +137,38 @@ - (void)layoutSublayers
             playbackSessionModel->setSpatialTrackingLabel(m_spatialTrackingLabel);
             playbackSessionModel->setSoundStageSize(WebCore::AudioSessionSoundStageSize::Large);
+
+                if (auto* playbackSessionModel = playbackSessionInterface().playbackSessionModel())
```
indentation is wrong.

jyavenard commented on Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.mm:
```diff
@@ -133,21 +137,38 @@ - (void)layoutSublayers
     playbackSessionInterface().stopObservingNowPlayingMetadata();
     [linearMediaPlayer() exitFullscreenWithCompletionHandler:makeBlockPtr([this, protectedThis = Ref { *this }, completionHandler = WTFMove(completionHandler)] (BOOL success, NSError *error) {
-        if (auto* playbackSessionModel = this->playbackSessionModel()) {
+        if (auto* playbackSessionModel = playbackSessionInterface().playbackSessionModel()) {
```
why this change? what does this add?

jyavenard commented on Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.mm:
```diff
@@ -199,7 +220,7 @@ - (void)layoutSublayers
     playbackSessionInterface().stopObservingNowPlayingMetadata();
     [linearMediaPlayer() exitExternalPlaybackWithCompletionHandler:makeBlockPtr([this, protectedThis = Ref { *this }, handler = WTFMove(exitHandler)] (BOOL success, NSError *error) mutable {
-        if (auto* playbackSessionModel = this->playbackSessionModel()) {
+        if (auto* playbackSessionModel = playbackSessionInterface().playbackSessionModel()) {
```
ditto. Why the change?

jyavenard commented on Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.mm:
```diff
@@ -169,7 +190,7 @@ - (void)layoutSublayers
     // Puts the player into `enteringExternal` state.
     [linearMediaPlayer() enterExternalPlaybackWithCompletionHandler:makeBlockPtr([this, protectedThis = Ref { *this }, handler = WTFMove(enterHandler)] (BOOL success, NSError *error) mutable {
-        if (auto* playbackSessionModel = this->playbackSessionModel()) {
+        if (auto* playbackSessionModel = playbackSessionInterface().playbackSessionModel()) {
```
ditto

jyavenard commented on Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.h:
```diff
@@ -189,6 +189,9 @@ class PlaybackSessionModelContext final
     WebCore::AudioSessionSoundStageSize soundStageSize() const final { return m_soundStageSize; }
     void setSoundStageSize(WebCore::AudioSessionSoundStageSize) final;
 
+        bool prefersAutoDimming() const final { return m_prefersAutoDimming; }
```
identation is wrong

jyavenard commented on Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.mm:
```diff
@@ -1155,6 +1155,7 @@ - (BOOL)prefersStatusBarHidden
     Ref interface = ensureInterface(contextId);
     interface->enterFullscreen();
 
+
```
why the extra line?

jyavenard commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -1194,7 +1248,7 @@ - (void)beganEnterFullScreenWithInitialFrame:(CGRect)initialFrame finalFrame:(CG
 #endif
 
-            if (auto* videoPresentationManager = self._videoPresentationManager) {
+            if (RefPtr videoPresentationManager = [self _videoPresentationManager]) {
```
can you take this in a follow-up patch? As mentioned in earlier reviews. Let's keep the scope narrow

Phinny01 replied:
@aprotyas  mentioned to change it in an earlier patch but I will revert and do them in another patch

jyavenard commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -1858,6 +1924,8 @@ - (BOOL)prefersSceneDimming
     return YES;
 }
 
+#if PLATFORM(VISION)
```
that seems like unrelated changes and isn't described in the commit log

jyavenard commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -1940,10 +2010,16 @@ - (void)_performSpatialFullScreenTransition:(BOOL)enter completionHandler:(Compl
-        || (!enter && stage.preferredDarkness != originalState.preferredDarkness)) {
+    MRUIDarknessPreference targetDarkness;
+
+    if (enter)
```
could be a one liner for the lot

jyavenard reviewed (commented):
LGTM, but you can't land something deleting the Xcode WebCore project file!

jyavenard commented on Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.mm:
```diff
@@ -136,18 +140,35 @@ - (void)layoutSublayers
             playbackSessionModel->setSpatialTrackingLabel(m_spatialTrackingLabel);
             playbackSessionModel->setSoundStageSize(WebCore::AudioSessionSoundStageSize::Large);
+
+            if (auto* playbackSessionModel = playbackSessionInterface().playbackSessionModel())
```
Why? you have already `retrieved playbackSessionModel` three lines above. why fetch it again?

Phinny01 replied:
I have removed the redundant retrieval now

jyavenard commented on Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.mm:
```diff
@@ -136,18 +140,34 @@ - (void)layoutSublayers
             playbackSessionModel->setSpatialTrackingLabel(nullString());
             playbackSessionModel->setSoundStageSize(WebCore::AudioSessionSoundStageSize::Automatic);
+
+            m_savedAutoDimmingPreference = playableViewController().prefersAutoDimming;
```
Is this even needed?

can prefersAutoDimming change ? I don't think we should store it here, considering we set it ourselves earlier, as such it serves no purpose.

Phinny01 replied:
The prefersAutoDimmingChanged method is actually not needed so I removed that function entirely to make it simpler and left that line as it is to just set m_savedAutoDimmingPreference once when leaving fullscreen.

aestes commented on Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.mm:
```diff
@@ -86,6 +88,8 @@ - (void)layoutSublayers
     : VideoPresentationInterfaceIOS { playbackSessionInterface }
 {
+    if (RefPtr model = videoPresentationModel())
+        model->addClient(*this);
```
Why is this necessary? The interface should already be a client of the model (via the call to `setVideoPresentationModel` in `VideoPresentationManagerProxy::createModelAndInterface`).

aestes commented on Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.mm:
```diff
@@ -76,6 +76,8 @@ - (void)layoutSublayers
 VideoPresentationInterfaceLMK::~VideoPresentationInterfaceLMK()
 {
+    if (RefPtr model = videoPresentationModel())
+        model->removeClient(*this);
```
This should already be handled by `VideoPresentationManagerProxy::invalidateInterface`.

aestes commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.h:
```diff
@@ -60,6 +60,10 @@
+- (void)setPrefersAutoDimming:(BOOL)prefersAutoDimming;
+
+- (void)didEnterFullscreen;
+- (void)didExitFullscreen;
```
I think these names are confusing, because both element fullscreen and video fullscreen have the concept of "entering/exiting fullscreen".

Perhaps `did(Enter|Exit)VideoFullscreen` are better names.

aestes commented on Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.h:
```diff
@@ -97,6 +97,8 @@ class VideoPresentationInterfaceLMK final : public WebCore::VideoPresentationInt
 
     CompletionHandler<void(bool)> m_exitExternalPlaybackHandler;
+
+    BOOL m_savedAutoDimmingPreference { YES };
```
From reading the code I don't think `m_savedAutoDimmingPreference` is necessary. I think you can remove it. Am I missing something?

aestes commented on Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.mm:
```diff
@@ -1121,6 +1121,40 @@
+    model->setPrefersAutoDimming(prefersAutoDimming);
+}
+
+void PlaybackSessionModelContext::setPrefersAutoDimming(bool value)
```
To keep things organized I'd move this function to where the other `PlaybackSessionModelContext` functions are defined.

aestes commented on Source/WebCore/platform/cocoa/PlaybackSessionModel.h:
```diff
@@ -150,6 +150,9 @@ class PlaybackSessionModel : public CanMakeWeakPtr<PlaybackSessionModel> {
     virtual bool supportsLinearMediaPlayer() const { return false; }
 #endif
 
+    virtual bool prefersAutoDimming() const { return true; }
```
Nit: I think `false` is a better default value here.

aestes commented on Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.h:
```diff
@@ -238,6 +241,7 @@ class PlaybackSessionModelContext final
     std::optional<WebCore::SpatialVideoMetadata> m_spatialVideoMetadata;
     std::optional<WebCore::VideoProjectionMetadata> m_videoProjectionMetadata;
 
+    bool m_prefersAutoDimming { true };
```
Nit: we should initialize this to `false`.

aestes commented on Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.mm:
```diff
@@ -1121,6 +1121,40 @@
+
+    auto it = m_contextMap.find(*m_controlsManagerContextId);
+    if (it == m_contextMap.end())
+        return true;
```
Nit: `return false`.

aestes commented on Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.mm:
```diff
@@ -1121,6 +1121,40 @@
+bool PlaybackSessionManagerProxy::prefersAutoDimming() const
+{
+    if (!m_controlsManagerContextId)
+        return true;
```
Nit: `return false`

aestes commented on Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.mm:
```diff
@@ -1121,6 +1121,40 @@
+void PlaybackSessionModelContext::setPrefersAutoDimming(bool value)
+{
+    if (m_prefersAutoDimming == value)
+        return;
```
Nit: this isn't necessary.

aestes commented on Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.mm:
```diff
@@ -1435,6 +1435,14 @@ - (BOOL)prefersStatusBarHidden
+#if PLATFORM(VISION)
+    if (RefPtr pageClient = page->pageClient()) {
+        if (auto fullScreenWindowController = pageClient->fullScreenWindowController()) {
+            if ([fullScreenWindowController respondsToSelector:@selector(didExitFullscreen)])
```
You can remove this `-respondsToSelector:`

aestes commented on Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.mm:
```diff
@@ -1456,6 +1464,14 @@ - (BOOL)prefersStatusBarHidden
+#if PLATFORM(VISION)
+    if (RefPtr pageClient = page->pageClient()) {
+        if (auto fullScreenWindowController = pageClient->fullScreenWindowController()) {
+            if ([fullScreenWindowController respondsToSelector:@selector(didEnterFullscreen)])
```
Ditto.

aestes commented on Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.mm:
```diff
@@ -1456,6 +1464,14 @@ - (BOOL)prefersStatusBarHidden
 #endif
     page->didEnterFullscreen(contextId);
+#if PLATFORM(VISION)
+    if (RefPtr pageClient = page->pageClient()) {
```
I would move this code into `PageClientImplIOS::didEnterFullscreen`.

aestes commented on Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.mm:
```diff
@@ -1435,6 +1435,14 @@ - (BOOL)prefersStatusBarHidden
     page->didExitFullscreen(contextId);
 
+#if PLATFORM(VISION)
+    if (RefPtr pageClient = page->pageClient()) {
```
I would move this code into `PageClientImplIOS::didExitFullscreen`.

Phinny01 replied:
I have done that now

aestes commented on Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.mm:
```diff
@@ -1435,6 +1435,14 @@ - (BOOL)prefersStatusBarHidden
 
+#if PLATFORM(VISION)
+    if (RefPtr pageClient = page->pageClient()) {
+        if (auto fullScreenWindowController = pageClient->fullScreenWindowController()) {
```
The null check is unnecessary since it's safe to send an ObjC message to nil.

aestes commented on Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.mm:
```diff
@@ -1456,6 +1464,14 @@ - (BOOL)prefersStatusBarHidden
     page->didEnterFullscreen(contextId);
+#if PLATFORM(VISION)
+    if (RefPtr pageClient = page->pageClient()) {
+        if (auto fullScreenWindowController = pageClient->fullScreenWindowController()) {
```
The null check is unnecessary since it's safe to send an ObjC message to nil.

aestes commented on Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
```diff
@@ -244,7 +244,9 @@ class PageClientImpl final : public PageClientImplCocoa
     WebFullScreenManagerProxyClient& fullScreenManagerProxyClient() final;
-
+#if PLATFORM(VISION)
+    WKFullScreenWindowController* fullScreenWindowController() override;
```
If you moved the didEnter/ExitFullscreen code into `PageClientImplIOS` then you could remove this.

aestes commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -872,6 +884,22 @@ - (void)resetSupportedOrientations
     [_fullscreenViewController resetSupportedOrientations];
 }
 
+- (void)setPrefersAutoDimming:(BOOL)prefersAutoDimming
```
It doesn't look like anything calls this.

aestes commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -920,6 +948,33 @@ - (void)enterFullScreen:(CGSize)mediaDimensions completionHandler:(CompletionHan
+            UIApplication.sharedApplication.mrui_activeStage.preferredDarkness = targetDarkness;
+        } else
+            UIApplication.sharedApplication.mrui_activeStage.preferredDarkness = MRUIDarknessPreferenceUnspecified;
+    }
```
```
if (!self.isFullScreen || !WebKit::useSpatialFullScreenTransition())
    return;

bool prefersAutoDimming = false;
if (RefPtr videoPresentationManager = [self _videoPresentationManager]) {
    if (RefPtr bestVideo = videoPresentationManager->bestVideoForElementFullscreen())
        prefersAutoDimming = bestVideo->playbackSessionModel()->prefersAutoDimming();
}
UIApplication.sharedApplication.mrui_activeStage.preferredDarkness = prefersAutoDimming ? MRUIDarknessPreferenceDark : MRUIDarknessPreferenceUnspecified;
```

aestes commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -1849,6 +1911,11 @@ - (BOOL)_sceneAspectRatioLockingEnabled
+    if (RefPtr videoPresentationManager = [self _videoPresentationManager]) {
+        if (RefPtr bestVideo = videoPresentationManager->bestVideoForElementFullscreen())
+            return bestVideo->playbackSessionModel()->prefersAutoDimming();
+    }
```
Let's move this below the `[self _sceneDimmingEnabled]` check.

aestes commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -1412,6 +1467,13 @@ - (void)_completedExitFullScreen:(CompletionHandler<void()>&&)completionHandler
+        MRUIStage *stage = UIApplication.sharedApplication.mrui_activeStage;
+        stage.preferredDarkness = [_parentWindowState preferredDarkness];
+    }
+#endif
```
This should not be necessary, because we already reset `preferredDarkness` to the `_parentWindowState` in `-_performSpatialFullScreenTransition:completionHandler:`.

Phinny01 replied:
That's right. I have removed this

aestes commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -1940,10 +2007,13 @@ - (void)_performSpatialFullScreenTransition:(BOOL)enter completionHandler:(Compl
-        || (!enter && stage.preferredDarkness != originalState.preferredDarkness)) {
+    MRUIDarknessPreference targetDarkness;
+
+    targetDarkness = enter ? (self.prefersSceneDimming ? MRUIDarknessPreferenceDark : originalState.preferredDarkness) : originalState.preferredDarkness;
```
Nit: you can combine these two lines.

aestes commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -2021,11 +2091,18 @@ - (void)toggleSceneDimming
+        if (RefPtr bestVideo = videoPresentationManager->bestVideoForElementFullscreen())
+            bestVideo->playbackSessionModel()->setPrefersAutoDimming(updatedPrefersSceneDimming);
+    } else
+        [[NSUserDefaults standardUserDefaults] setBool:updatedPrefersSceneDimming forKey:kPrefersFullScreenDimmingKey];
```
It seems wrong to not update the user default when there is a `bestVideo`. As we discussed in person, let's move the getting/setting of the `kPrefersFullScreenDimmingKey` to the `PlaybackSessionModel`.

Phinny01 replied:
I have moved the getting/setting to the PlaybackSessionModel

aestes reviewed (changes requested):
I think we can still simplify this quite a bit. Since `PlaybackSessionModel` is the source of truth for `prefersAutoDimming`, we should move the code that reads/writes to user defaults to `PlaybackSessionModel` (i.e. it should be constructed with a value for `m_prefersAutoDimming` set from the user default, and `PlaybackSessionModel::setPrefersAutoDimming()` should update the user default). I'd move all the user default code out of `WKFullScreenWindowController`.

aestes commented on Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.mm:
```diff
@@ -63,6 +64,7 @@
     : m_manager(manager)
     , m_contextId(contextId)
 {
+    m_prefersAutoDimming = [[NSUserDefaults standardUserDefaults] boolForKey:@"WebKitPrefersFullScreenDimming"];
```
Nit: this could be moved into the initializer list
```suggestion
    , m_prefersAutoDimming([NSUserDefaults standardUserDefaults] boolForKey:@"WebKitPrefersFullScreenDimming"])
{
```

aestes commented on Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.mm:
```diff
@@ -333,6 +335,14 @@
 
+void PlaybackSessionModelContext::setPrefersAutoDimming(bool value)
+{
+    m_prefersAutoDimming = value;
```
I would first check that `m_prefersAutoDimming != value` before setting user defaults (which is a little expensive).

aestes commented on Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.mm:
```diff
@@ -333,6 +335,14 @@
+{
+    m_prefersAutoDimming = value;
+
+    if (RefPtr manager = m_manager.get())
```
Why do you need to check if `m_manager` is non-null?

Phinny01 replied:
we don't need it anymore. it was an oversight but I have removed it now

aestes commented on Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.mm:
```diff
@@ -1121,6 +1131,34 @@
+bool PlaybackSessionManagerProxy::prefersAutoDimming() const
+{
+    if (!m_controlsManagerContextId)
+        return [[NSUserDefaults standardUserDefaults] boolForKey:@"WebKitPrefersFullScreenDimming"];
```
```suggestion
        return false;
```

Phinny01 replied:
done

aestes commented on Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.mm:
```diff
@@ -1121,6 +1131,34 @@
+
+    auto it = m_contextMap.find(*m_controlsManagerContextId);
+    if (it == m_contextMap.end())
+        return [[NSUserDefaults standardUserDefaults] boolForKey:@"WebKitPrefersFullScreenDimming"];
```
```suggestion
        return false;
```

aestes commented on Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.mm:
```diff
@@ -1121,6 +1131,34 @@
+
+void PlaybackSessionManagerProxy::setPrefersAutoDimming(bool prefersAutoDimming)
+{
+    [[NSUserDefaults standardUserDefaults] setBool:prefersAutoDimming forKey:@"WebKitPrefersFullScreenDimming"];
```
Let's remove this line and rely on `PlaybackSessionModelContext::setPrefersAutoDimming` to do this.

aestes commented on Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
```diff
@@ -348,8 +348,8 @@ class PageClientImpl final : public PageClientImplCocoa
-    void didEnterFullscreen() final { };
-    void didExitFullscreen() final { };
+    void didEnterFullscreen() override;
+    void didExitFullscreen() override;
```
Nit: you should still use `final` here.

aestes commented on Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
```diff
@@ -896,6 +895,20 @@ static UIInterfaceOrientationMask toUIInterfaceOrientationMask(WebCore::ScreenOr
 
+void PageClientImpl::didEnterFullscreen()
+{
+#if ENABLE(VIDEO_PRESENTATION_MODE) && PLATFORM(VISION)
```
I don't think you need the `PLATFORM(VISION)` part here.

Phinny01 replied:
I removed it

aestes commented on Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
```diff
@@ -896,6 +895,20 @@ static UIInterfaceOrientationMask toUIInterfaceOrientationMask(WebCore::ScreenOr
+
+void PageClientImpl::didExitFullscreen()
+{
+#if ENABLE(VIDEO_PRESENTATION_MODE) && PLATFORM(VISION)
```
I don't think you need the `PLATFORM(VISION)` part here.

aestes commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -1852,8 +1886,10 @@ - (BOOL)prefersSceneDimming
+            return bestVideo->playbackSessionModel()->prefersAutoDimming();
+    }
 
     return YES;
```
I think we should return `NO` in the default case.

aestes commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -790,6 +793,8 @@ @implementation WKFullScreenWindowController {
     BOOL _enterRequested;
     BOOL _exitingFullScreen;
 
+    BOOL _prefersAutoDimming;
```
I think we can remove this ivar completely. It will fall out of sync when `-videoControlsManagerDidChange` is called, and we never actually read from it anyway.

Phinny01 replied:
Right! I removed it

aestes commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -812,6 +817,12 @@ - (id)initWithWebView:(WKWebView *)webView
+        if (RefPtr bestVideo = videoPresentationManager->bestVideoForElementFullscreen())
+            _prefersAutoDimming = bestVideo->playbackSessionModel()->prefersAutoDimming();
+    }
+#endif
```
We can remove this.

Phinny01 replied:
I removed that block of code

aestes commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -2021,11 +2058,17 @@ - (void)toggleSceneDimming
+            bestVideo->playbackSessionModel()->setPrefersAutoDimming(updatedPrefersSceneDimming);
+    }
+
+    _prefersAutoDimming = updatedPrefersSceneDimming;
```
We can remove this line.

aestes commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -2021,11 +2058,17 @@ - (void)toggleSceneDimming
+    _prefersAutoDimming = updatedPrefersSceneDimming;
 
     if (self.isFullScreen) {
+        MRUIDarknessPreference target =updatedPrefersSceneDimming ? MRUIDarknessPreferenceDark: (_parentWindowState ? [_parentWindowState preferredDarkness]: MRUIDarknessPreferenceUnspecified);
```
```suggestion
        MRUIDarknessPreference target = updatedPrefersSceneDimming ? MRUIDarknessPreferenceDark : (_parentWindowState ? [_parentWindowState preferredDarkness] : MRUIDarknessPreferenceUnspecified);
```

Phinny01 replied:
addressed this in the new patch

aestes reviewed (approved):
looks great! r=me with some comments.

Commit:
Fixed Auto-dimming effect in portal window of APMP playback from WebView's Div Element Fullscreen
https://bugs.webkit.org/show_bug.cgi?id=296495
rdar://154223902

Reviewed by Andy Estes and Jean-Yves Avenard.

This patch creates a single place to store a user preference: whether videos should automatically
dim the background when played in fullscreen. PlaybackSessionModel carries a boolean
m_prefersAutoDimming, with accessors that notify its clients when the value changes.
PlaybackSessionModelContext forwards the change to VideoPresentationManagerProxy, allowing every
concurrent playback context to observe the current setting. PlaybackSessionManagerProxy persists
the value to user defaults and relays it to the Web-process fullscreen manager so the system knows
which darkness level to apply. VideoPresentationInterfaceLMK applies the flag to the underlying
AVPlayerViewController, caches the previous stage darkness on entry and restores it on exit.
WKFullScreenWindowControllerIOS consults the proxy when entering fullscreen, reflects user changes
made regarding autoDimming back into the proxy.

* Source/WebCore/platform/cocoa/PlaybackSessionModel.h:
(WebCore::PlaybackSessionModel::prefersAutoDimming const):
(WebCore::PlaybackSessionModel::setPrefersAutoDimming):
* Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.mm:
(WebKit::VideoPresentationInterfaceLMK::presentFullscreen):
(WebKit::VideoPresentationInterfaceLMK::dismissFullscreen):
(WebKit::VideoPresentationInterfaceLMK::ensurePlayableViewController):
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.mm:
(WebKit::PlaybackSessionModelContext::PlaybackSessionModelContext):
(WebKit::PlaybackSessionModelContext::setPrefersAutoDimming):
(WebKit::PlaybackSessionManagerProxy::prefersAutoDimming const):
(WebKit::PlaybackSessionManagerProxy::setPrefersAutoDimming):
* Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.mm:
(WebKit::VideoPresentationManagerProxy::didExitFullscreen):
(WebKit::VideoPresentationManagerProxy::didEnterFullscreen):
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/WebFullScreenManagerProxy.h:
(WebKit::WebFullScreenManagerProxy::client):
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::fullScreenManagerProxyClient):
(WebKit::PageClientImpl::didEnterFullscreen):
(WebKit::PageClientImpl::didExitFullscreen):
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.h:
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
(-[WKFullScreenWindowController initWithWebView:]):
(-[WKFullScreenWindowController didEnterVideoFullscreen]):
(-[WKFullScreenWindowController didExitVideoFullscreen]):
(-[WKFullScreenWindowController prefersSceneDimming]):
(-[WKFullScreenWindowController _performSpatialFullScreenTransition:completionHandler:]):
(-[WKFullScreenWindowController toggleSceneDimming]):
* Source/WebKit/WebKitSwift/LinearMediaKit/LinearMediaTypes.swift:
(WKSPlayableViewControllerHost.prefersAutoDimming):
(WKSPlayableViewControllerHost.playable):
* Source/WebKit/WebKitSwift/LinearMediaKit/WKSLinearMediaTypes.h:

Canonical link: https://commits.webkit.org/298720@main
